Implementing Chess

Priwinn opened the pull request:
-Implements (western) chess
-Special moves: castling and en passant
-Includes edge case rules: Draw by repetition (can be disabled), Half move role
-Implements a drag and drop GUI

hopshackle commented on .vscode/settings.json:
```diff
@@ -0,0 +1,5 @@
+{
+    "java.compile.nullAnalysis.mode": "disabled",
+    "java.configuration.updateBuildConfiguration": "interactive",
+    "java.debug.settings.onBuildFailureProceed": true
+}
```
You should exclude the .vscode directory from the repository (include in the .gitignore file). this is local to your setup only.

hopshackle commented on src/main/java/core/Game.java:
```diff
@@ -845,34 +845,34 @@ public String toString() {
-//        players.add(new BasicMCTSPlayer());
+        players.add(new RandomPlayer());
+    //    players.add(new RandomPlayer());
+    //    players.add(new BasicMCTSPlayer());
 //        players.add(new OSLAPlayer());
 //        players.add(new RMHCPlayer());
 
-        RMHCParams params = new RMHCParams();
-        params.horizon = 15;
-        params.discountFactor = 0.99;
-        params.heuristic = AbstractGameState::getHeuristicScore;
+        // RMHCParams params = new RMHCParams();
+        // params.horizon = 15;
+        // params.discountFactor = 0.99;
+        // params.heuristic = AbstractGameState::getHeuristicScore;
 //        AbstractPlayer rmhcPlayer = new RMHCPlayer(params);
 //        players.add(rmhcPlayer);
 
-//        MCTSParams mcts_params = new MCTSParams();
-//        players.add(new MCTSPlayer(mcts_params));
+       MCTSParams mcts_params = new MCTSParams();
+       players.add(new MCTSPlayer(mcts_params));
 
-        players.add(new HumanGUIPlayer(ac));
-        players.add(new HumanGUIPlayer(ac));
-        players.add(new HumanGUIPlayer(ac));
-//        players.add(new HumanConsolePlayer());
+        // players.add(new HumanGUIPlayer(ac));
+        // players.add(new HumanGUIPlayer(ac));
+        // players.add(new HumanGUIPlayer(ac));
+    //    players.add(new HumanConsolePlayer());
 //        players.add(new FirstActionPlayer());
 
```
Not a big deal, but it is better practice not to include any temporary testing changes as in Game.main() in the final PR.

hopshackle commented on src/main/java/games/chess/actions/EnPassant.java:
```diff
@@ -0,0 +1,136 @@
+import games.chess.components.ChessPiece;
+import java.lang.System;
+import java.util.Objects;
+
+/**
+ * <p>Actions are unit things players can do in the game (e.g. play a card, move a pawn, roll dice, attack etc.).</p>
+ * <p>Actions in the game can (and should, if applicable) extend one of the other existing actions, in package {@link core.actions}.
+ * Or, a game may simply reuse one of the existing core actions.</p>
+ * <p>Actions may have parameters, so as not to duplicate actions for the same type of functionality,
+ * e.g. playing card of different types (see {@link games.sushigo.actions.ChooseCard} action from SushiGo as an example).
+ * Include these parameters in the class constructor.</p>
+ * <p>They need to extend at a minimum the {@link AbstractAction} super class and implement the {@link AbstractAction#execute(AbstractGameState)} method.
+ * This is where the main functionality of the action should be inserted, which modifies the given game state appropriately (e.g. if the action is to play a card,
+ * then the card will be moved from the player's hand to the discard pile, and the card's effect will be applied).</p>
+ * <p>They also need to include {@link Object#equals(Object)} and {@link Object#hashCode()} methods.</p>
+ * <p>They <b>MUST NOT</b> keep references to game components. Instead, store the {@link Component#getComponentID()}
+ * in variables for any components that must be referenced in the action. Then, in the execute() function,
+ * use the {@link AbstractGameState#getComponentById(int)} function to retrieve the actual reference to the component,
+ * given your componentID.</p>
+ */
```
This boiler plate makes sense in the gametemplate package - but when you implement concrete classes you shoudl replace it with doc that makes sense to the implementation (i.e. here, one or two lines about En Passant...or even just delete it to avoid duplicated cruft)

hopshackle commented on src/main/java/games/chess/ChessGameState.java:
```diff
@@ -0,0 +1,414 @@
+    }
+
+    public void deletePiece(ChessPiece piece) {
+        if (piece == null) {
```
This is potentially dangerous as it can easily lead to bugs happening invisibly and being very difficult to track down. 
If code calls deletePiece() then they are expecting a piece to be present...if it isn't then a much better approach is to throw an Error (IllegalArgument, say).
This will highlight the bug as early as possible and make it much easier to fix

hopshackle commented on src/main/java/games/chess/actions/MovePiece.java:
```diff
@@ -0,0 +1,111 @@
+        ChessGameState gs = (ChessGameState) ags;
+        ChessPiece piece = gs.getPiece(startX, startY);
+
+        //Delete the piece in the start position
+        gs.deletePiece(piece); // Remove the piece from its original position
+
```
What happens if there is no piece there? We should throw an error or do something?

hopshackle commented on src/main/java/games/chess/actions/Promotion.java:
```diff
@@ -0,0 +1,79 @@
+    @Override
+    public boolean execute(AbstractGameState gs) {
+        MovePiece move = new MovePiece(startX, startY, targetX, targetY);
+        if (!move.execute(gs)) {
+            return false;
+        }
```
In this case you should definitely throw an error - or better, do this within the move.execute() method. 
If the move fails then something unexpected has happened, and it is dangerous to let this fail silently and carry on as if everything was fine.

hopshackle commented on src/main/java/games/chess/components/ChessPiece.java:
```diff
@@ -0,0 +1,137 @@
+    //Wether the piece has moved or not (e.g. for castling, en passant, etc.), only relevant for pawns, kings and rooks
+    private MovedState movedState;
+
+    private int x;
+    private int y;
```
You are storing the position of a piece on the ChessPiece, and also recording the ChessPiece on an (x, y) position on the ChessBoard.

Both of these are individually very reasonable, but you have the same information stored in two separate places, which opens up all sorts of possible subtle bugs if the two sets of information get out of synch with each other.

This may be a refactor too far, but consider if there is a way to have a single source of the information. For example, could you make ChessPiece immutable and just have the mutable information stored on the chessboard.

hopshackle replied:
Perfectly reasonable conclusion.

hopshackle commented on src/main/java/games/chess/ChessGameState.java:
```diff
@@ -0,0 +1,414 @@
+    @Override
+    protected ChessGameState _copy(int playerId) {
+        ChessGameState copy = new ChessGameState(getGameParameters(), getNPlayers());
+        copy.whitePieces = new ArrayList<>();
+        copy.whitePieces.addAll(whitePieces.stream().map(ChessPiece::copy).toList());
```
You could replace this I think with copy.whitePieces = new ArrayList(whitePieces); which is much more concise and avoids streams

hopshackle replied:
Yes - my mistake...I forgot ChessPiece is mutable.

hopshackle commented on src/main/java/games/chess/ChessGameState.java:
```diff
@@ -0,0 +1,414 @@
+
+    @Override
+    public double getGameScore(int playerId) {
+        return 0;
```
This should check for a terminal game state, in which case the score is 0, 0.5, 1 for loss/draw/win?

hopshackle commented on src/main/java/games/chess/ChessParameters.java:
```diff
@@ -0,0 +1,54 @@
+
+public class ChessParameters extends TunableParameters {
+
+    public int maxRounds = 300; // Maximum number of moves in the game before it ends in a draw
+    public int drawByRepetition = 3; // Number of times a position must be repeated for a draw by repetition to be declared, 0 for no limit
```
This is hiding a parameter of the same name in the grandparent AbstractParameters. 

You can use the existing functionality here - just set maxRounds to be the value required.

Priwinn replied:
Changing this to new name because using the existing functionality messes with the order of checks (checks maxRounds before checkmate)

hopshackle commented on src/test/java/games/chess/ActionTests.java:
```diff
@@ -0,0 +1,205 @@
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+import java.util.List;
```
Hurrah! Unit tests.

hopshackle commented:
I pressed submit on the PR review without summarising:
Main changes needed are:
1) Remove the private .vscode directory from the PR...this shoudl not go anywhere near the public repository.
2) Remove the duplicated comments that have been cut and paste from gametemplate (either replace them with something meaningful and useful for chess, or delete them to avoid cluttering the screen)
3) Throw errors when something unexpected happens; don't just fail silently. This is a quick route to very buggy and unreliable code.

Priwinn commented:
Made a bit of a mess rebasing to remove vscode and Game.java changes.
Added a couple of exception throws.
Removed remaining template text.
Minor fixes.

hopshackle commented:
Not sure if you're finished - but when I check out this branch, I have major compilation errors in ChessParameters, EnPassant and Castle due to unresolved merges?

Also - main remaining question is why have a separate maxChessRounds at all - just use the existing maxRounds?

Priwinn commented:
Sorry, not sure what happened but not it should compile correctly, also thanks to an anonymous playtester I found a bug for the bishops so that is fixed as well. I changed back to using the existing implementation of maxRounds.
